ratis-server-api/src/main/java/org/apache/ratis/server/protocol/TermIndex.java

```diff
@@ -60,43 +72,46 @@ static TermIndex valueOf(LogEntryProto proto) {
 
   /** @return a {@link TermIndex} object. */
   static TermIndex valueOf(long term, long index) {
-    return new TermIndex() {
-      @Override
-      public long getTerm() {
-        return term;
-      }
+    TermIndexKey key = new TermIndexKey(term, index);
+    try {
+      return CACHE.get(key, () -> new TermIndexImpl(term, index));
+    } catch (ExecutionException e) {
+      throw new IllegalStateException("Failed to valueOf(" + term + ", " + index + ")", e);
+    }
+  }
 
-      @Override
-      public long getIndex() {
-        return index;
-      }
+  class TermIndexKey {
+    private final long term;
+    private final long index;
 
-      @Override
-      public boolean equals(Object obj) {
-        if (obj == this) {
-          return true;
-        } else if (!(obj instanceof TermIndex)) {
-          return false;
-        }
-
-        final TermIndex that = (TermIndex) obj;
-        return this.getTerm() == that.getTerm()
-            && this.getIndex() == that.getIndex();
-      }
+    public TermIndexKey(long term, long index) {
+      this.term = term;
+      this.index = index;
+    }
 
-      @Override
-      public int hashCode() {
-        return Long.hashCode(term) ^ Long.hashCode(index);
-      }
+    public long getTerm() {
+      return term;
+    }
 
-      private String longToString(long n) {
-        return n >= 0L? String.valueOf(n) : "~";
-      }
+    public long getIndex() {
+      return index;
+    }
 
-      @Override
-      public String toString() {
-        return String.format("(t:%s, i:%s)", longToString(term), longToString(index));
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
       }
-    };
+      TermIndexKey that = (TermIndexKey) o;
+      return term == that.term && index == that.index;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(term, index);
```

Comment:
Let's don't change the `hashCode()` and keep `toString()`; otherwise it becomes an incompatible change.